Library naming conventions

MaartenBent opened the pull request:
Following the discussion in [#19164](https://trac.wxwidgets.org/ticket/19164), I updated the library naming conventions document.

I think it mostly covers the names when using configure (Windows, Linux, and I assume macOS too), and the names when using the VS solutions and makefile.gcc/vc on Windows.
I didn't include the `CFG` option (yet,) it only affects the output directory. It could be included in the `setup.h` section.

I think in some cases CMake generated projects slightly differ from this convention (flavour/vendor/host related). I'll fix that in this PR a well.

MaartenBent committed:
Update library naming convention documentation

PBfordev commented:
My nitpick suggestions:
1. Provide examples of actual names of a library and a _setup.h_ path, at least for one platform (MSW?).
2. Perhaps mention that the convention applies to the default multilib build. I have no idea whether monolithic build is supported outside Windows.

vadz commented on docs/contributing/about-platform-toolkit-and-library-names.md:
```diff
@@ -22,61 +21,97 @@ for wxWidgets, but is encompassing enough to maintain a relatively complete
+
+--------------------------------------------------------------------
+
+`$debug` is set to `'d'` for debug libraries and is empty for release libraries.
+It is only used in libraries created with MSVC and gcc/vc makefile. On UNIX,
+it is always empty.
```
```suggestion
`$debug` is set to `'d'` for the libraries using debug version of CRT and is empty
for release libraries. It is only really useful for the libraries created with MSVC
projects and makefiles, as MSVC debug and release CRT are not ABI-compatible,
but is also used by `makefile.gcc` under MSW for consistency with `makefile.vc`. 
When using configure under MSW or UNIX, it is always empty.
```

vadz reviewed (approved):
Thanks for updating this, but I think this actually belong to the manual and not under `docs/contributing`. Of course, it can be moved there as a separate step later.

I'm also still not sure what does CMake do...

MaartenBent committed:
CMake: Improve library names

Use the same output names as MSVC and makefile.gcc/vc on Windows, and the same as configure on other platforms.

MaartenBent committed:
CMake: Create symlinks with so version

MaartenBent committed:
CMake: Use a macro to get the flavour

Also include the flavour in builtin target names and include path.

MaartenBent committed:
CMake: Improve wxrc name

Include the version and flavour in the target name, no need to rename anymore.
Add the wxrc symlink to extra uninstall files, not the actual target name.

MaartenBent committed:
CMake: Improve webkit plugin name

MaartenBent commented:
> * Provide examples of actual names of a library and a _setup.h_ path, at least for one platform (MSW?).

Since this document is for wx developers, not for wx users, I don't know if this will be useful.

> * Perhaps mention that the convention applies to the default multilib build. I have no idea whether monolithic build is supported outside Windows.

It also applies to monolithic build on all platforms. `mono` will the `$name` (instead of for example `core` or `aui`).

---

The PR is ready. 

CMake now generates the same output as MSVC and makefile.gcc/vc on Windows, and the same output as configure on other platforms. I tested shared and static builds with flavours and vendors and the installed files (bin/include/lib) are all the same.

@PBfordev In the mailing list you mentioned compatibility with 3rd party tools such as Code::Blocks and Codelite. Are you able to test if this has been improved?

If there is a demand for it, I could make the naming configurable (override `WIN32_MSVC_NAMING`), and add an option similar to the `CFG` option in config.gcc/vc.

PBfordev commented:
> CMake now generates the same output as MSVC and makefile.gcc/vc on Windows, and the same output as configure on other platforms. I tested shared and static builds with flavours and vendors and the installed files (bin/include/lib) are all the same.

I have just tested (sorry for the delay) and it seems that the names of GCC import libraries built with CMake generator "MinGW Makefiles" still do not exactly match those produced with bundled GCC makefiles. CMake ones have an extra ".dll" there., e.g.,  _libwxbase31ud_**.dll**_.a_ vs _libwxbase31ud.a_.
 
> @PBfordev In the mailing list you mentioned compatibility with 3rd party tools such as Code::Blocks and Codelite. Are you able to test if this has been improved?

AFAICT, if the libraries names produced with CMake are the same as with makefiles, the compatibility will greatly improve. The issue with these tools is that they still expect wxWidgets to use the old in-source build idiom, where all the necessary files (headers and libraries) are inside WXDIR folder with clearly defined structure, i.e., they do not work with out-of-the source builds. However, at least when the library names are same, the libraries can be just copied to the their WXDIR folders. 

Additionally, IIRC, CodeLite's wx-config-win for some reason also [needs file _build.cfg_](https://github.com/eranif/codelite/issues/2659) (in build's folder mswu or mswud folder), which is created when using makefiles but not built with CMake nor shipped with prebuilt binaries.

To conclude, when it comes to the third party tools, I believe that just having the same names for mingw CMake and makefile is good enough and as far as wxWidgets should go.

MaartenBent commented:
> I have just tested (sorry for the delay) and it seems that the names of GCC import libraries built with CMake generator "MinGW Makefiles" still do not exactly match those produced with bundled GCC makefiles. CMake ones have an extra ".dll" there., e.g., _libwxbase31ud_**.dll**_.a_ vs _libwxbase31ud.a_.

Thanks for testing. I forgot to mention this, the extension is indeed different. I wasn't sure if this would be a problem, but it should be simple to fix using [SUFFIX](https://cmake.org/cmake/help/latest/prop_tgt/SUFFIX.html).

> Additionally, IIRC, CodeLite's wx-config-win for some reason also [needs file _build.cfg_](https://github.com/eranif/codelite/issues/2659) (in build's folder mswu or mswud folder), which is created when using makefiles but not built with CMake nor shipped with prebuilt binaries.

I'll have a look at this file and see what I can do in CMake. Looks like it contains the used values of `config.gcc`.

PBfordev commented:
> Thanks for testing. I forgot to mention this, the extension is indeed different. I wasn't sure if this would be a problem, but it should be simple to fix using [SUFFIX](https://cmake.org/cmake/help/latest/prop_tgt/SUFFIX.html).

This affects also CMake's FindwxWidgets() module, I am not sure what you mean by "fix", you mean by the user?


> I'll have a look at this file and see what I can do in CMake. Looks like it contains the used values of `config.gcc`.

It also contains wxWidgets version but as I wrote, I would not bother with this.

MaartenBent commented:
> This affects also CMake's FindwxWidgets() module, I am not sure what you mean by "fix", you mean by the user?

No, I mean I will fix it.

PBfordev commented:
> Thanks for testing. I forgot to mention this, the extension is indeed different. I wasn't sure if this would be a problem, but it should be simple to fix using [SUFFIX](https://cmake.org/cmake/help/latest/prop_tgt/SUFFIX.html).

Sorry, I forgot to write that while I know basically nothing about CMake, but to change the suffix for the import libraries, I modified [IMPORT_SUFFIX](https://cmake.org/cmake/help/latest/prop_tgt/IMPORT_SUFFIX.html) in my wxAutoExcel:
https://github.com/PBfordev/wxAutoExcel/blob/master/CMakeLists.txt#L179

MaartenBent committed:
CMake: Use .a suffix for import libraries

So it will have the same extension as the libraries created by makefile.gcc.

MaartenBent committed:
CMake: generate build.cfg

MaartenBent commented:
> [IMPORT_SUFFIX](https://cmake.org/cmake/help/latest/prop_tgt/IMPORT_SUFFIX.html)

Yes, that is the correct property.

I think this PR is now ready to be merged.

PBfordev commented on docs/contributing/about-platform-toolkit-and-library-names.md:
```diff
@@ -22,61 +21,99 @@ for wxWidgets, but is encompassing enough to maintain a relatively complete
 
-For MSW (native hosted only) libraries the library name should be of
-the form:
+*Windows libraries* (using VS solution or makefile.gcc/vc):
```
Perhaps mention that it applies also for CMake when the generator is any of the Visual Studio or NMake and MinGW Makefiles?

MaartenBent replied:
Good point. I forgot to mention CMake in this document.

MaartenBent committed:
Mention CMake in library naming convention documentation

vadz commented:
Thanks for the fixes and thanks @PBfordev for the review! I admit I haven't actulaly reviewed the CMake part myself, but the changes to the document look correct to me, so I'm just going to merge this as you both think it's good.

Thanks again!
